compiler/rustc_lint/src/transmute.rs

```diff
@@ -94,6 +129,81 @@ impl<'tcx> LateLintPass<'tcx> for CheckTransmutes {
 
         check_ptr_transmute_in_const(cx, expr, body_owner_def_id, const_context, src, dst);
         check_unnecessary_transmute(cx, expr, callee, arg, const_context, src, dst);
+        check_int_to_ptr_transmute(cx, expr, arg, src, dst);
+    }
+}
+
+/// Check for transmutes from integer to pointers (*const/*mut, &/&mut and fn()).
+///
+/// Using the resulting pointers would be undefined behavior.
+fn check_int_to_ptr_transmute<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx hir::Expr<'tcx>,
+    arg: &'tcx hir::Expr<'tcx>,
+    src: Ty<'tcx>,
+    dst: Ty<'tcx>,
+) {
+    if matches!(src.kind(), ty::Uint(_) | ty::Int(_))
+        && let ty::Ref(_, inner_ty, mutbl) | ty::RawPtr(inner_ty, mutbl) = dst.kind()
+        // bail-out if the argument is literal 0 as we have other lints for those cases
+        && !matches!(arg.kind, hir::ExprKind::Lit(hir::Lit { node: LitKind::Int(v, _), .. }) if v == 0)
+        // bail-out if the inner type if a ZST
+        && cx.tcx
+            .layout_of(cx.typing_env().as_query_input(*inner_ty))
+            .is_ok_and(|layout| !layout.is_1zst())
+    {
+        // does the argument needs parenthesis
+        let mut paren_left = "";
+        let mut paren_right = "";
+        if matches!(arg.kind, hir::ExprKind::Binary(..)) {
+            paren_left = "(";
+            paren_right = ")";
+        }
+
+        cx.tcx.emit_node_span_lint(
+            INTEGER_TO_PTR_TRANSMUTES,
+            expr.hir_id,
+            expr.span,
+            IntegerToPtrTransmutes {
+                suggestion: if !cx.tcx.hir_is_inside_const_context(expr.hir_id) {
```

Comment:
```suggestion
                // FIXME: once https://github.com/rust-lang/rust/issues/144538 finishes,
                // we can always recommend the method.
                suggestion: if !cx.tcx.hir_is_inside_const_context(expr.hir_id) {
```

Reply:
I already had added a similar comment (as requested earlier), but in the `lints.rs` file. Added that yours as well.